### PR TITLE
Remove structural match from `TypeId`

### PR DESCRIPTION
**Pull request by raldone01**
As per https://github.com/rust-lang/rust/pull/99189#issuecomment-1203720442.

> Removing the structural equality might make sense, but is a breaking change that'd require a libs-api FCP.

https://github.com/rust-lang/rust/pull/99189#issuecomment-1197545482

> Landing this PR now (well, mainly the "remove structural equality" part) would unblock `const fn` `TypeId::of`, since we only postponed that because we were guaranteeing too much.

See also #99189, #101698

### CODE REVIEW

**Comment by thomcc** [comment]
I can't start that FCP and the comment is from @m-ou-se so....

r? @m-ou-se

**Comment by chriss0612** [comment]
cc @m-ou-se @eddyb

File: src/test/ui/consts/issue-73976-monomorphic.rs
Comment (eddyb) [inline src/test/ui/consts/issue-73976-monomorphic.rs:24]: Should this be fixed before landing?

**Review by eddyb — APPROVED** [review]
LGTM but I don't understand what's going on with the test.

File: src/test/ui/consts/issue-73976-monomorphic.rs
Comment (eddyb) [reply src/test/ui/consts/issue-73976-monomorphic.rs:24]: Oh, is that all handled by this?
* #101698

Comment (chriss0612) [reply src/test/ui/consts/issue-73976-monomorphic.rs:24]: Yes with only this PR and not https://github.com/rust-lang/rust/pull/101698 it is impossible to do this test.

**Comment by eddyb** [comment]
Marking as blocked on:
* https://github.com/rust-lang/rust/pull/101698

**Comment by WaffleLapkin** [comment]
Is this waiting on libs FCP to be started?

**Comment by m-ou-se** [comment]
This removes the [StructuralPartialEq](https://doc.rust-lang.org/stable/std/marker/trait.StructuralPartialEq.html) implementation from TypeId, by removing derive(PartialEq) and replacing it with a manual `impl PartialEq for TypeId`.

@rfcbot merge

**Comment by dtolnay** [comment]
Could someone give an example of what code this breaks? I found the following but it only works currently on nightly. If this is the only thing, then I think it would be fine to merge this without T-libs-api FCP.

```rust
#![feature(const_type_id)]

use std::any::TypeId;

const INT: TypeId = TypeId::of::<i32>();

fn main() {
    match TypeId::of::<u8>() {
        INT => println!("i32"),
        _ => println!("not i32"),
    }
}
```

```console
error: to use a constant of type `TypeId` in a pattern, `TypeId` must be annotated with `#[derive(PartialEq, Eq)]`
 --> src/main.rs:9:9
  |
9 |         INT => println!("i32"),
  |         ^^^
```

**Comment by coolreader18** [comment]
Theoretically something like this would break, right?

```rust
trait A {
    const ID: TypeId;
}
fn foo<T: A>(id: TypeId) {
    match id {
        T::ID => {}
        _ => {}
    }
}
```
but obviously you wouldn't be able to actually impl such a trait on stable, so.

File: library/core/src/any.rs
Comment (a1phyr) [inline library/core/src/any.rs:677]:
```suggestion
    #[inline]
    fn eq(&self, other: &Self) -> bool {
        self.t == other.t
    }
```

This produces less IR and allows the compiler to "see through" this function

**Comment by dtolnay** [comment]
> Theoretically something like this would break, right?

As far as I can tell, code like that would not break, because it has never compiled in the past.

```console
error[E0158]: associated consts cannot be referenced in patterns
 --> src/lib.rs:6:9
  |
6 |         T::ID => {}
  |         ^^^^^
```

**Review by dtolnay — CHANGES_REQUESTED** [review]
@raldone01 this needs a rebase to resolve conflict with #110393. Since you created the PR from an org fork instead of a personal one, Rust maintainers do not have permission to update the PR for you.

7fd34b19e48bbb0d226d63926b7dfc7ae293e7ab

**Comment by raldone01** [comment]
The following tests fail: `x test --stage 1 tests/ui/consts/issue-73976-monomorphic.rs tests/ui/consts/const_cmp_type_id.rs tests/ui/const-generics/generic_const_exprs/typeid-equality-by-subtyping.rs`

I am not use how to resolve these.

If structural match and `const_cmp_type_id` are both removed there won't be any way to compare type ids in const contexts.

**Comment by dtolnay** [comment]
The commit I linked in https://github.com/rust-lang/rust/pull/103291#pullrequestreview-1444228962 has the test fix.

`x.py test tests/ui --bless` is usually sufficient to update those tests for a standard library change.

**Review by dtolnay — APPROVED** [review]
Thanks!

**Comment by dtolnay** [comment]
The pinnacle of efficiency. This must be a record?

Screenshot from 2023-05-26 13-06-38

**Comment by jackh726** [comment]
sorry to necro this a bit - looking through our I-unsound known-bug tracking in #105107 and came across this.

> @dtolnay I added a commit that blesses the tests, but test tests/ui/const-generics/generic_const_exprs/typeid-equality-by-subtyping.rs uses Structural match of TypeId, so i removed the `// run-pass` and added #110395 as known-bug since comparing TypeId in const context is now imposible.
> 
> I'm not sure if this is how this should be done, since there was already a I-unsound related known bug on this test.

For future reference: this is not the correct use of `known-bug`. #110395 is not a _bug_ for one. This obviously was resolved over time, but in the future please tag the appropriate team (in this case T-types) to check what the appropriate course is for handling changes to known-bug cases (you could imagine, for example, that an alternative resolution would be that #97156 could have been *closed* with this PR).

**Pull request merged**
